### PR TITLE
Add multi-language system prompts and BedrockChatAdapter implementation

### PR DESCRIPTION
**Pull request by michel-heon**
### **Pull Request: Centralize and Internationalize System Prompts**

This pull request addresses the issue of scattered system prompts across the codebase and the lack of support for internationalization, as described in the corresponding Git issue [https://github.com/aws-samples/aws-genai-llm-chatbot/issues/571](https://github.com/aws-samples/aws-genai-llm-chatbot/issues/571).
 
---

### **Changes:**
- **Commit:** `256279db811d17f6c558ccf469bfbec0e0d93583`
- **Objective:** Centralize system prompt management and enable internationalization (i18n).
- **Affected Files:**
  - `system_prompts.py`: A new module created to centralize all system prompts and support multiple languages (English and Canadian French).
  - `base.py`: Refactored methods (`get_prompt`, `get_condense_question_prompt`, `get_qa_prompt`) to retrieve prompts from `system_prompts.py`.
  - `__init__.py`: Updated to import system prompts for simplified access.
  
---

### **Key Improvements:**
- **Centralization:** All system prompts are now stored in one place (`system_prompts.py`), improving manageability and scalability.
- **Internationalization:** The system can now support multiple languages, with English and Canadian French prompts available as an initial setup. This framework allows for future expansion into other languages.
- **Adapter Compatibility:** Adapters such as `azure-openai`, `mistral`, `claude`, `titan`, and `llama` are updated to use the new prompt management system, ensuring consistency and reducing code duplication.

---

### **Testing Instructions:**
1. **Deploy the Application:** Follow the deployment instructions provided in the project documentation to deploy the application.
2. **Target Language:** The language used by the system prompts is now managed by the mechanism in `system_prompts.py`.
3. **Modify Language:** You can modify the target language at runtime by editing the language selection in the `GenAIChatBotStack-LangchainInterfaceReques` Lambda function.
4. **Prompt Tracing:** The system prompts are logged and can be traced in CloudWatch. Additionally, they are available in the `prompt` field of the metadata variable in the AWS GenAI Chatbot console for further analysis.

---

### **Expected Outcome:**
- Simplified prompt management across adapters.
- Enhanced scalability and flexibility in adding new languages.

---

### CODE REVIEW

**Review by charles-marion — COMMENTED** [review]
Thank you for creation this PR! (And sorry for the delay)

The build process failed. I recommend to run `npm run-vetall`

File: lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py
Comment (michel-heon) [reply lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py:151]: My observation is that this prompt is a default prompt provided by langchain. It applies to models that do not have an adapter in the library. In fact, it's important to internationalize this variable too, to standardize the linguistic context of the conversation. For English, the behavior doesn't change, but for French, there's a gain in conversational quality.

For the moment, I haven't made any changes to this code segment.

File: lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py
Comment (michel-heon) [reply lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py:40]: In fact... very, very honestly, this system-prompt was generated by chatgpt itself. It's a case of meta-prompting (a prompt that generated the prompt). In short, you'd have to do a test to see if the response would perform better with 'context' rather than 'document'. In any case, this system-prompt has enabled us to obtain excellent results, especially in terms of the quality of containment in the knowledge domain provided, which seems to me to be an essential issue in RAG.

No change in this code segment for the moment.

Translated with www.DeepL.com/Translator (free version)

File: lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py
Comment (michel-heon) [reply lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py:39]: Yes, especially when it comes to limiting the field of knowledge. In fact, this prompt gives even better results than the English one. Again, this is a case of meta-prompting by chatgpt.

No change in this code segment for the moment.

File: lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py
Comment (charles-marion) [reply lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py:151]: I agree with you. However this is a change of behavior since this use case would have a different prompt. (I can test it at some point)

File: lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py
Comment (michel-heon) [reply lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py:15]: I'm all for thinking things through, as both solutions offer their own list of advantages and disadvantages. I propose to create a new issue for this feature after this PR is closed.

File: lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py
Comment (michel-heon) [reply lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py:19]: I agree, This will be changed

File: lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py
Comment (michel-heon) [reply lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py:25]: I agree, This will be changed

File: lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py
Comment (michel-heon) [reply lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py:109]: I agree, This will be changed

File: lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py
Comment (michel-heon) [reply lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py:45]: Sorry, I misunderstood. I've changed the prompt to reflect the comment.

File: lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py
Comment (michel-heon) [reply lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py:27]: I agree

**Comment by charles-marion** [comment]
@michel-heon  please tag me or click the re-request review button if you'd like me to have a look.
(Note I noticed the file `base.py_new` which is probably not relevant?)

**Comment by michel-heon** [comment]
@charles-marion In fact, the base.py_new file is an error that I've deleted. And indeed, the code is ready for revision.

File: lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py
Comment (charles-marion) [inline lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py:262]: Changing the JSON format here would break the metric in the dashboard. Please undo
https://github.com/aws-samples/aws-genai-llm-chatbot/blob/main/lib/monitoring/index.ts#L289

File: lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py
Comment (charles-marion) [inline lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py:44]: This is only applicable for bedrock. Why did you add it here?

File: lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py
Comment (charles-marion) [inline lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py:367]: I think this is a copy of https://github.com/aws-samples/aws-genai-llm-chatbot/blob/main/lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py

It would revert this change in this file.

File: lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py
Comment (charles-marion) [inline lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py:31]:
```suggestion
    logger.debug("No guardrails ID found.")
```
Otherwise it will be logged on every llm call.

File: lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py
Comment (charles-marion) [inline lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py:103]: It would not set the value if temperature is 0
```suggestion
        if temperature is not None:
```

File: lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py
Comment (charles-marion) [inline lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py:178]:
```suggestion
```

File: lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py
Comment (charles-marion) [inline lib/model-interfaces/langchain/functions/request-handler/adapters/bedrock/base.py:181]:
```suggestion
        logger.debug(f"condense_question_prompt: {condense_question_prompt}")
```
I would recommend to mark them all as debug to reduce cloudwatch usage. (nitpick sorry)

File: lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py
Comment (charles-marion) [inline lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py:37]:
```suggestion
```
I would remove this because there is already a global log level setting here
https://github.com/aws-samples/aws-genai-llm-chatbot/blob/main/lib/shared/index.ts#L52

**Review by charles-marion — CHANGES_REQUESTED** [review]
Thank you for the update.
Note that I appreciate the help with this change. I can address my comments and complete the PR if you'd like (I don't want to take too much of your time)

File: lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py
Comment (michel-heon) [reply lib/model-interfaces/langchain/functions/request-handler/adapters/base/base.py]: Is it possible to include this information in the developer's guide documentation?

File: lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py
Comment (charles-marion) [inline lib/model-interfaces/langchain/functions/request-handler/adapters/shared/prompts/system_prompts.py:76]: @michel-heon I was verifying the change before merging and I noticed this prompt in english and french mention <conv> which is not part of the prompt structure (the chat history is not inside of <conv>).

I was considering reverting/rewording to what it was before.

Do you have details about the change? 

For reference this is the previous version
```
Given the following conversation and a follow up question, rephrase the follow up question to be a standalone question.
```

**Comment by michel-heon** [comment]
The i18n mechanism works for Bedrock, but not for azureopenai. This fix could be part of a future PR?

**Comment by charles-marion** [comment]
> The i18n mechanism works for Bedrock, but not for azureopenai. This fix could be part of a future PR?

No we cannot since it will break any current usage of it.
Is it a quick fix? Note this PR is not urgent to be merged since someone could check out the PR if they need translation.

Also please note my other question: https://github.com/aws-samples/aws-genai-llm-chatbot/pull/576#discussion_r1815512269

**Comment by michel-heon** [comment]
In fact, the use of `azureopenai` refers to the default system-prompt in `langchain`, which must be overloaded. The same applies to Mistral. I agree with deferring this development to another PR and merging the current work.

**Comment by charles-marion** [comment]
Build is blocked until the following is merged. #598 
I will merge this PR once it is reviewed and merged.

**Review by charles-marion — APPROVED** [review]
I ran the integration tests and fixed the formatting.

LGTM. Thank you for you contribution!

**Pull request merged**
